### ISSUE TITLE
Calling onPause()/onResume() on the gvrLayout causes problems when not preserving EGL context

### ISSUE THREAD

**Issue by Consti10** [googlevr/gvr-android-sdk#558]
Until now I was able to use gvrLayout with a SurfaceView that does not preserve the egl context on Pause by simply not calling onResume/inPause on the layout. This introduced another bug: [https://github.com/googlevr/gvr-android-sdk/issues/556](url)  @jdduke why I cannot use that approach any more.

So the problem when calling `setPreserveEGLContextOnPause(false);`  on the view inserted into the gvrLayout is that this results in 3 unnecessary creations and deletions on the EGL context. Only the 3rd creating of the egl context seems to be succesfull and the applications "onDraw()" loop is called successive.

When logging the egl context creations in GLSurfaceViewEGL14 (a derivate of GLSurfaceView using EGL14) the logcat looks like in the comment below.

Or shortened into the essential:
`egl createSurface
onEGLContextCreated()
onSurfaceSizeChanged()
onEGLContextDestroyed()

egl createSurface
onEGLContextCreated()
onSurfaceSizeChanged()
onEGLContextDestroyed()

egl createSurface
onEGLContextCreated()
onSurfaceSizeChanged()
onEGLContextDestroyed()`

The problem: This is not the logcat when manually pausing / resuming the activity, but rather just opening it & letting it run. E.g. called in StereoActivity are only `onCreate()` and 'onResume' once, but the result is like when the activity is paused and resumed 3 times in total

I thought that maybe the EGL Surface is first created with false dimensions when placed inside the gvrLayout, but  setting the size manually using mGLView14Stereo.getHolder().setFixedSize(2560,1440);
didn't change anything.

The problem occured on both of my testing devices (ZTE axon 7, Nexus 5X) and while I didn't count the number of context creations/deletions on the Nexus 5X It looked like 3 EGL context creations / deletions, too. (on the ZTE it is 3 creations/deletions, like in the logcat above)

Any ideas how to fix that, e.g. not preserving the egl context onPause() and also letting GLSurfaceView create the EGL context only once ?

**Comment by Consti10** [comment]
10136-10136 V: BoostFramework() : mPerf = com.qualcomm.qti.Performance@d6e8ba7
    
    --------- beginning of system
10136-10136 D: fontPath: 
10136-10136 W: commandlineflags.cc:1503 Ignoring RegisterValidateFunction() for flag pointer 0x7f6ec1f688: no flag found at that address
10136-10136 D: Fetched params from VrParamsProvider: allow_dynamic_java_library_loading: true
    allow_dynamic_library_loading: true
    allow_vrcore_compositing: false
    allow_vrcore_head_tracking: false
    async_reprojection_config <
      additional_ahardwarebuffer_usage: 268435456
      back_rgb16_with_bgr16: true
      black_boost: 0
      display_latency_micros: 900
      flags: 16
      strips_per_frame: 2
      vsync_grace_period_micros: 0
    >
    cpu_late_latching_enabled: false
    daydream_image_alignment: 3
    dim_ui_layer: true
    disallow_multiview: false
    enable_forced_tracking_compat: false
    screen_capture_config <
      allow_casting: false
      allow_screen_record: false
      allow_screenshot: false
    >
    touch_overlay_enabled: false
    use_device_idle_detection: true
    use_direct_mode_sensors: false
    use_magnetometer_in_sensor_fusion: true
    use_online_magnetometer_calibration: false
    use_stationary_bias_correction: false
    use_system_clock_for_sensor_timestamps: false
10136-10136 D: ignored Vulkan layer search path /data/app/com.google.vr.vrcore-1/lib/arm64:/data/app/com.google.vr.vrcore-1/base.apk!/lib/arm64-v8a for namespace 0x7f88810160
10136-10136 W: getBasePackageName in Asset, index=3, return 'com.google.vr.vrcore'
10136-10136 W: commandlineflags.cc:1503 Ignoring RegisterValidateFunction() for flag pointer 0x7f69e38828: no flag found at that address
10136-10136 I: [vr/gvr/capi/src/gvr_core_api_loader_android.cc:182] Successfully loaded GVR library version 1.140.0 from VrCore (target was 1.140.0).
10136-10136 I: [vr/gvr/capi/src/gvr.cc:103] Initialized GVR version 1.140.0
10136-10193 D: No Network Security Config specified, using platform default
10136-10193 I: RegisterTcmMonitor from: com.android.okhttp.TcmIdleTimerMonitor
10136-10136 I: Successfully loaded GvrLayout from VrCore.
10136-10136 I: GLThreadManager instance created
10136-10136 V: BoostFramework() : mPerf = com.qualcomm.qti.Performance@536b66d
    BoostFramework() : mPerf = com.qualcomm.qti.Performance@92f12a2
10136-10202 I: Noticed surfaceView surface lost tid=4158
10136-10136 I: Repro enabled:false
10136-10136 D: fontPath: 
10136-10136 I: onResume tid=4158
10136-10136 I: Repro enabled:false
10136-10182 E: void android::uirenderer::renderthread::CanvasContext::setSurface(android::Surface *) to create EGLSurface constantin.fpv_vr.wifibroadcast/constantin.fpv_vr.activities.Activity_Stereo
10136-10202 I: Noticed surfaceView surface acquired tid=4158
10136-10202 D: MSAA Level:4
10136-10202 I: Noticing that we want render notification tid=4158
10136-10202 W: egl createSurface
10136-10202 D: GLSurfaceViewEGL14 MODE:MODE_NORMAL
10136-10202 W: onEGLContextCreated()
10136-10202 V: Priority set to -16 in GLRendererStereo
10136-10207 V: Priority set to -16 in VideoPlayerN VideoReceiver
10136-10207 V: SetSocketTimeout error
10136-10208 V: Priority set to -2 in TelRN WB
10136-10208 V: SetSocketTimeout error
10136-10202 W: onSurfaceSizeChanged()
10136-10202 V: already stopped
    circularRefreshT stopped
    circularRefreshT started
10136-10209 V: Priority set to -2 in TextElements
10136-10202 I: Sending render notification tid=4158
10136-10136 I: onPause tid=4158
    onPause waiting for mPaused.
10136-10202 I: mPaused is now true. tid=4158
    Releasing EGL surface because paused tid=4158
10136-10202 I: Releasing EGL context because paused tid=4158
10136-10202 W: onEGLContextDestroyed()
10136-10202 I: GL14 del video start
10136-10202 I: GL14 del video stop
    GL14 release start
    GL14 release stop
10136-10202 V: circularRefreshT stopped
10136-10136 D: fontPath: 
10136-10136 I: onResume tid=4158
10136-10136 I: onResume waiting for !mPaused.
10136-10202 I: mPaused is now false. tid=4158
10136-10202 D: MSAA Level:4
10136-10202 W: egl createSurface
10136-10202 D: GLSurfaceViewEGL14 MODE:MODE_NORMAL
10136-10202 W: onEGLContextCreated()
10136-10202 V: Priority set to -16 in GLRendererStereo
10136-10212 V: Priority set to -16 in VideoPlayerN VideoReceiver
10136-10212 V: SetSocketTimeout error
10136-10213 V: Priority set to -2 in TelRN WB
10136-10213 V: SetSocketTimeout error
10136-10136 I: Repro enabled:false
10136-10136 I: onPause tid=4158
    onPause waiting for mPaused.
10136-10202 W: onSurfaceSizeChanged()
10136-10202 V: already stopped
    circularRefreshT stopped
    circularRefreshT started
10136-10214 V: Priority set to -2 in TextElements
10136-10202 I: mPaused is now true. tid=4158
    Releasing EGL surface because paused tid=4158
10136-10202 I: Releasing EGL context because paused tid=4158
10136-10202 W: onEGLContextDestroyed()
10136-10202 I: GL14 del video start
10136-10202 I: GL14 del video stop
10136-10202 I: GL14 release start
    GL14 release stop
10136-10202 V: circularRefreshT stopped
10136-10136 D: fontPath: 
10136-10136 I: onResume tid=4158
10136-10136 I: onResume waiting for !mPaused.
10136-10202 I: mPaused is now false. tid=4158
10136-10202 D: MSAA Level:4
10136-10202 W: egl createSurface
10136-10202 D: GLSurfaceViewEGL14 MODE:MODE_NORMAL
10136-10202 W: onEGLContextCreated()
10136-10136 I: Repro enabled:false
10136-10215 V: Priority set to -16 in VideoPlayerN VideoReceiver
10136-10202 V: Priority set to -16 in GLRendererStereo
10136-10215 V: SetSocketTimeout error
10136-10216 V: Priority set to -2 in TelRN WB
10136-10216 V: SetSocketTimeout error
10136-10182 D: endAllActiveAnimators on 0x7f6dfb0800 (RippleDrawable) with handle 0x7f6dfb19c0
10136-10202 W: onSurfaceSizeChanged()
10136-10202 V: already stopped
    circularRefreshT stopped
    circularRefreshT started
10136-10220 V: Priority set to -2 in TextElements
10136-10202 V: OpenGL FPS:54.617676
10136-10202 V: OpenGL FPS:59.850374
10136-10202 V: OpenGL FPS:59.410884
10136-10202 V: OpenGL FPS:59.970015
10136-10202 V: OpenGL FPS:59.880240 ...

**Comment by Consti10** [comment]
I continued debugging around this one issue and came to the root of the problem (I assume) :
Calling onPause/onResume on the GvrLayout results in 2 unnecessary changes in the activity state.

E.g while not interacting with the phone in any way after opening the activity via a button on resume() in the activity is actually called 3 times !

**Comment by jdduke** [comment]
Hi @Consti10, for the case where the Activity is being resumed multiple times, this can happen on Daydream-ready phones (like the ZTE Axon 7) when the VR session is first started. Have you enabled the "Skip VR entry screens" in the developer options, by chance?

**Comment by Consti10** [comment]
I investigated the n of onResume() calls depending on "skip vr entries" and which headset !!! is set
Skip vr entry screens not enabled:
Cardboard headset: onResume() called once (how it should be)
Daydream view headset: I am unable to use a daydream headset with vertex displacement distortion correction, because the "entry screen" doesn't allow me ("incompatible app: This cardboard application is not compatible with daydream headsets"). Well, it actually is, but how can I tell gvr that ? 

Skip VR entry screens enabled:
Cardboard headset:  onResume() called once (how it should be) again
Daydream headset: onResume() called 3 times

**Comment by Consti10** [comment]
So to solve the issue I just needed to be able to tell gvr that my app is compatible with a daydream headset (as long as you don't mind the cromatic aberation, it does the distortion correction also for daydream headsets, and has lower video latency than when using async reprojection)

**Comment by jdduke** [comment]
Vertex distortion is incompatible with async reprojection, as you noted, but we require async reprojection for all Daydream apps (https://developers.google.com/vr/distribute/daydream/performance-requirements#PS-P2). Why are you more concerned about video latency than motion-to-photon latency? You won't be able to get the necessary ~20ms motion-to-photon latency on a Daydream-ready phone without frontbuffer rendering, which is implied with GVR's async reprojection.

To indicate that your app is Daydream-compatible, you must specify that in the Activity manifest using `<category android:name="com.google.intent.category.DAYDREAM" />`. See also https://developers.google.com/vr/reference/vr-manifest#platform_compatibility.

**Comment by Consti10** [comment]
I see. Wouldn't actually clarify my app as a "Daydream app" but I would like my users to be able to use it with a "Daydream headset". Currently you have to enable the "skip vr entry screens" to use my (and probably also all cardboard apps) with a daydream headset.  It would be nice to have some java gvrApi call to basically say "hey, I am not an daydream app but would like to use the daydream headset as a normal VR headset.

My app is for FPV (first person view) aerial drone flight. There latency of the video feed matters the most. 
 Therefore I developed a custom front buffer renderer that supports vertex displacement distortion correction & has <20ms input-to-output latency.
That was actually 2 years ago. Back then there was no support for a video texture running on the async reprojection thread in Daydream.

**Comment by Consti10** [comment]
For reference: only adding the above didn't change anything, I also had to add `<action android:name="android.intent.action.MAIN" />` to my activity. e.g. 
`<activity
            android:name=".activities.AStereo"
            android:screenOrientation="landscape">
            <intent-filter>
                <action android:name="android.intent.action.MAIN" />
                <category android:name="com.google.intent.category.DAYDREAM" />
            </intent-filter>
        </activity>`

This is a bit in contrast with my Apps workflow, which manages settings, establishing a connection to the aircraft usw in 2D (normal Android App UI)  and only enters VR via a button if requested.
So the user shouldn't be able to start the VR activity from "Daydream home" but with these settings he can.

But now I am stuck at the "controller pairing" screen ?

The problem I am seeing here is that this activity is not a "Daydream activity" because it breaks with the basic daydream principles like "always have a controller for navigation, usw" , it just wants to use the daydream headset as a normal VR headset.

So going down this route might not be ideal and might introduce more problems. Any other idea ?

**Comment by jdduke** [comment]
> So the user shouldn't be able to start the VR activity from "Daydream home" but with these settings he can.

You can prevent the Activity from appearing in the Daydream VR launcher by marking it with `android:exported="false"`. 

`The problem I am seeing here is that this activity is not a "Daydream activity" because it breaks with the basic daydream principles like "always have a controller for navigation, usw" , it just wants to use the daydream headset as a normal VR headset.`

Does your app not use the Cardboard trigger at all? We're considering adding support for a Daydream app declaring "optional" controller support, which would avoid the controller connection UI when entering VR. Would that work for you?

Note that that would still interrupt your Activity to show the 2D transition screen.

Going back to the root of the problem, is the brief Activity resume->pause->resume sequence causing problems besides redundant GL context creation? Or are you simply wanting to optimize the flow and avoid some of the associated jank?

**Comment by Consti10** [comment]
> You can prevent the Activity from appearing in the Daydream VR launcher by marking it with android:exported="false".

Thanks that worked as expected.

> Does your app not use the Cardboard trigger at all? We're considering adding support for a Daydream app declaring "optional" controller support, which would avoid the controller connection UI when entering VR. Would that work for you?

No, I'm not using it. That would be great. How long would it be until that's possible ?

> Note that that would still interrupt your Activity to show the 2D transition screen.
I can get rid of that using `mGvrLayout.getUiLayout().setTransitionViewEnabled(false);` 

No, it does not cause other problems than the redundant GL context creation.  (At least no problems that weren't my fault, I had tied the lifecycle of almost all my threads to the context creation/deletion and therefore experienced  multithreading issues. I already fixed that.).  Nonetheless, documenting this behaviour would be great. It took me quite a while to figure it out.

**Comment by jdduke** [comment]
> > Does your app not use the Cardboard trigger at all? We're considering adding support for a Daydream app declaring "optional" controller support, which would avoid the controller connection UI when entering VR. Would that work for you?
> No, I'm not using it. That would be great. How long will it be until that's possible ?

Potentially with the next release of Google VR Services (in ~6 weeks).

> I can get rid of that using mGvrLayout.getUiLayout().setTransitionViewEnabled(false);

That works for Cardboard, but for Daydream we require the transition view. We might revisit that decision in the future to reduce VR entry friction.

> Nonetheless, documenting this behaviour would be great. It took me quite a while to figure it out.

We'll soon be publishing some documentation for all of the developer options, including the "Skip VR entry screens" option. I'll make sure we make a note that this currently will result in a rapid resume/pause/resume sequence of the app on Daydream-ready devices.

**Comment by Consti10** [comment]
Looks good. I'l have an eye on it.
